### ISSUE TITLE
App icons / favicon should be configurable

### ISSUE THREAD

**Issue by jryans** [vector-im/riot-web#12020]
Riot admins may want to customise the various icons we have to use something else, like their own logo, etc. We should support some way of configuring this.

**Comment by fooness** [comment]
Would appreciate this very much. Specifying custom favicon and app icons of all flavor via e.g. `config.json` is a very important feature regarding customisation.

**Comment by t3chguy** [comment]
It won't be wholly possible for a lot of favicon consumers. Config.json is read and actioned in Javascript but a lot of things which use the favicon like scrapers don't run JS. for in browser we don't actually use all variants of the favicon anyway and just generate one using favico.js

**Comment by fooness** [comment]
Are there any other ways of achieving this without `config.json` but keeping everything upstream-compatible and without manual invasive changes?

**Comment by t3chguy** [comment]
Manually replacing the favicon files and then not overwriting them whenever you update?

**Comment by t3chguy** [comment]
![image](https://user-images.githubusercontent.com/2403652/79067460-5da25300-7cb7-11ea-98de-898f5b8f88e6.png)

just a few (37) to replace

**Comment by fooness** [comment]
> just a few (37) to replace

Whoops. And some of them seem to have arbitraty (random/hashed/whatever) additions to their names. Anyone knows how those files are automatically generated?

**Comment by t3chguy** [comment]
Only one is hashed, `favicon.4535157.ico`. Generated by https://github.com/vector-im/riot-web/blob/develop/webpack.config.js#L255-L289

**Comment by fooness** [comment]
In my folder, there’s many more. Also in your screenshot, highlighted below. Sorry if I’m wrong, not sure where those files come from and if they are used, but they definitely exist.

![79067460-5da25300-7cb7-11ea-98de-898f5b8f88e6](https://user-images.githubusercontent.com/11408027/79068288-0f488080-7cc6-11ea-85f4-cc2de52e18f2.png)

**Comment by t3chguy** [comment]
oh indeed, missed those when scanning through, so yeah that'll be webpack hashing them for cache-busting.

**Comment by fooness** [comment]
That means we’d have to manually re-name our image files each build iteration, I fear.

**Comment by t3chguy** [comment]
Indeed, but the only way around that would be introducing a backend to riot which serves the right assets hashed, which is unlikely to happen as riot is meant to be super simple to host anywhere without any special backend, just a bunch of static files

**Comment by t3chguy** [comment]
Actually, it could be done at the webserver level, forwarding `favicon.*.ico` -> your `some-other-folder/favicon.ico`

**Comment by fooness** [comment]
@t3chguy You mean like forwarding this via nginx or apache?

**Comment by t3chguy** [comment]
Well it wouldn't be "forwarding" as riot-web has to be served by some webserver anyhow. But yes, having Nginx handle requests for `favicon.<any_hash>.ico` with returning your own `some-other-folder/favicon.ico`

**Comment by zeigerpuppy** [comment]
A simple way to achieve this would be to have an `override` folder, in which the user can place their own favicons, background and logo.  If the folder is empty, then riot defaults to its normal directories.  This type of override survives updates as well.

**Comment by t3chguy** [comment]
But is not supported natively by HTML. You can't have an `<img src="..." fallback="...">` and even if you could, it'd make loading the images slower in the default (non-override) case, as you first have to load (and fail 404) the override before starting to load the original image.

**Comment by rtest12** [comment]
No one has new beautiful ideas?

**Comment by aceArt-GmbH** [comment]
> Well it wouldn't be "forwarding" as riot-web has to be served by some webserver anyhow. But yes, having Nginx/Apache handle requests for `favicon.<any_hash>.ico` with returning your own `some-other-folder/favicon.ico`

I came up with following nginx rule

```nginx
location ~* (/vector-icons/.*)\..*(\..*)$ {
    alias /src/images$1$2;
}
```
With the icons generated by https://github.com/vector-im/element-web/blob/develop/scripts/make-icons.sh mounted in `/src/images` ( `/src/images/vector-icons/...`).
You have to rename the images as they are named `android-chrome-144x144.png` while Element expects them as `android-chrome-144.png`

**Comment by aceArt-GmbH** [comment]
To change the opengraph image in a compose/kubernetes environment:
1. change the command to a custom script `command: /element-entry.sh`
1. mount it `- ./element-entry.sh:/element-entry.sh`
1. create the file
```sh
#!/bin/sh
sed -i -E 's/<meta property="og:image" content=".*" \/>/<meta property="og:image" content="https:\/\/chat.your.domain\/pub_img\/opengraph.jpg" \/>/' /app/index.html
# start normal command
nginx -g "daemon off;"
```

**Comment by strk** [comment]
> With the icons generated by https://github.com/vector-im/element-web/blob/develop/scripts/make-icons.sh

That script does't exist anymore there

**Comment by Domoel** [comment]
any updates regarding this? would be great if we could customize it.
